dm/config/source_config.go

```diff
@@ -353,3 +353,85 @@ func (c *SourceConfig) check(metaData *toml.MetaData, err error) error {
 	c.adjust()
 	return nil
 }
+
+// YamlForDowngrade returns YAML format represents of config for downgrade.
+func (c *SourceConfig) YamlForDowngrade() (string, error) {
+	s := NewSourceConfigForDowngrade(c)
+
+	// encrypt password
+	cipher, err := utils.Encrypt(utils.DecryptOrPlaintext(c.From.Password))
+	if err != nil {
+		return "", err
+	}
+	s.From.Password = cipher
+
+	// omit default values, so we can ignore them for later marshal
+	s.omitDefaultVals()
+
+	return s.Yaml()
+}
+
+// SourceConfigForDowngrade is the base configuration for source in v2.0.
+// This config is used for downgrade(config export) from a higher dmctl version.
+// When we add any new config item into SourceConfig, we should update it also.
+type SourceConfigForDowngrade struct {
+	EnableGTID      bool                   `yaml:"enable-gtid"`
+	AutoFixGTID     bool                   `yaml:"auto-fix-gtid"`
+	RelayDir        string                 `yaml:"relay-dir"`
+	MetaDir         string                 `yaml:"meta-dir"`
+	Flavor          string                 `yaml:"flavor"`
+	Charset         string                 `yaml:"charset"`
+	EnableRelay     bool                   `yaml:"enable-relay"`
+	RelayBinLogName string                 `yaml:"relay-binlog-name"`
+	RelayBinlogGTID string                 `yaml:"relay-binlog-gtid"`
+	UUIDSuffix      int                    `yaml:"-"`
+	SourceID        string                 `yaml:"source-id"`
+	From            DBConfig               `yaml:"from"`
+	Purge           PurgeConfig            `yaml:"purge"`
+	Checker         CheckerConfig          `yaml:"checker"`
+	ServerID        uint32                 `yaml:"server-id"`
+	Tracer          map[string]interface{} `yaml:"tracer"`
+	// any new config item, we mark it omitempty
+	CaseSensitive bool                  `yaml:"case-sensitive,omitempty"`
+	Filters       []*bf.BinlogEventRule `yaml:"filters,omitempty"`
+}
+
+// NewSourceConfigForDowngrade creates a new base config for downgrade.
+func NewSourceConfigForDowngrade(sourceCfg *SourceConfig) *SourceConfigForDowngrade {
+	return &SourceConfigForDowngrade{
+		EnableGTID:      sourceCfg.EnableGTID,
```

Comment:
maybe we can use reflect to copy these fields. I'm afraid we forget to copy newly added field here

Reply:
I add a clone func in 6fac04b , but I don't have confidence, so use it in unit_test.

Reply:
i think reflect is too magic and not easy for read and debug.. maybe we shold keep convert those struct manually